client/consensus/babe/src/authorship.rs

```diff
@@ -139,31 +141,42 @@ fn claim_secondary_slot(
 		*randomness,
 	)?;
 
-	for (pair, authority_index) in key_pairs {
-		if pair.public() == *expected_author {
+	for (authority_id, authority_index) in keys {
+		if authority_id == expected_author {
 			let pre_digest = if author_secondary_vrf {
-				let transcript = super::authorship::make_transcript(
+				let result = keystore.read().vrf_sign(
```

Comment:
nit: this API has way too many arguments and I'd prefer it to be extracted to a `VrfSigningRequest` struct which is the single parameter to the function.

Reply:
The result of `make_transcript` has the type `merlin::Transcript` which is very broad so I would expect only that the transcript needs to be passed to a signing request.

I guess that this is because the async signing request is meant to go to a remote signer that has more information on what it is signing. In that case, the API design presented in this PR does not accurately represent the goals or contract of the features you are intending to eventually deliver.

And without knowing the context of Substrate's roadmap, it's impossible to make any case in favor of this change.

I recommend starting again with better APIs that avoid leaking BABE consensus logic out to the signer. And if BABE data (not logic!) needs to be sent to a signing service, the API should be clear that the API is for BABE, not _any_ VRF signature as `vrf_sign` would imply.

Reply:
I totally agree that it wasn't a good design. I have now changed things so that the function gets `transcript` which is passed from BABE as suggested.

Reply:
hmm, after actually implementing it... i now realize that the there was actually a reason behind passing all those parameters to `vrf_sign`, mainly that the `transcript` needs to somehow be serialized so that it can be sent over the wire to the remote signer. However, doesn't seem that this is possible, which is why sending the parameters to the remote signer to build the transcript there was my go-to solution.

What do you think about this?

Reply:
I'm OK with that under these two conditions:
  - The interface is "honest". It only works for BABE, so it should be specific that it's about BABE. Best if this interface is on an extension trait to `Keystore` not the main `Keystore` traits as it ruins clean abstraction otherwise. 
  - The BABE code is also responsible for computing the transcript and checking whether the resulting VRF is within the primary threshold. It is not the responsibility of the signing service to do consensus-critical checks.

However I still find that approach far from ideal because it is probably hard to do the necessary trait-juggling to impose a clean boundary between Keystore & BABE. And we would be entering into this approach with the understanding that every consensus engine will now need its own extension trait for the Keystore. And it is a little harder to answer the question "how would a user implement their own consensus engine with raw signing?". IMO that is how we should view consensus crates and core code anyway - core interfaces should be general enough that altering consensus engine behavior doesn't require changing anything in the core. 

It would better to do a deeper refactoring so we can send the raw components of the transcript over the wire. Then you could have an actual `vrf_sign` function which accepted `transcript_components` in a general way. This would require refactoring the flow of how transcripts are created in BABE, which should be done carefully to ensure that the actual created transcripts are not any different. But in principle transcripts are just `merlin::Transcript::new(label)` and then a series of `transcript.append_message(label, message)`. 

maybe set up `vrf_sign` as a generic function like this:

```rust
fn vrf_sign<T: Encode>( // maybe `Serialize`. Don't worry about `Deserialize` though.
	&self,
	key_type: KeyTypeId,
	public: &Sr25519Public,
	transcript_data: T,
	make_transcript: impl Fn(&T) -> merlin::Transcript,
) -> ...
```

`transcript_data`, `key_type`, `public` can all go over the wire, even though the transcript doesn't. But the details of what the transcript data is and how it is transformed into a transcript are 100% up to the caller, as it should be.

Reply:
As an aside, I vastly prefer `sr25519_vrf_sign` as a name because we might add another VRF implementation to Substrate at some point and our API should accommodate that without requiring a major version bump.

Reply:
Thank you for such a detailed explanations. I tried to go with your suggested approach, but apparently this would cause the following issue:
```
error[E0038]: the trait `traits::BareCryptoStore` cannot be made into an object
   --> primitives/core/src/traits.rs:211:1
    |
64  |   pub trait BareCryptoStore: Send + Sync {
    |             --------------- this trait cannot be made into an object...
...
199 |       fn sr25519_vrf_sign<VRFData: Encode>(
    |          ---------------- ...because method `sr25519_vrf_sign` has generic type parameters
...
211 | / sp_externalities::decl_extension! {
212 | |     /// The keystore extension to register/retrieve from the externalities.
213 | |     pub struct KeystoreExt(BareCryptoStorePtr);
214 | | }
    | |_^ the trait `traits::BareCryptoStore` cannot be made into an object
    |
    = help: consider moving `sr25519_vrf_sign` to another trait
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

To build up on your idea, i will introduce `BabeAuthorshipVrfTranscriptData` but i will change the `vrf_sign` method interface to accept the data pre-encoded so that they are sent over the wire and on the call-site, the result would be decoded accordingly.

Reply:
The generic parameter issue is caused by both `<VRFData: Encode>` and the fact that one of method parameters is a generic (`make_transcript`)

Reply:
I have implemented an alternative that i think makes `sr25519_vrf_sign` generic enough. Could you please let me know your thoughts again?

Reply:
The issue is that it causes the trait to become non object-safe. I didn't realize that it was used as a trait object somewhere, which complicates things...